backend/src/datasource.js

```diff
@@ -0,0 +1,67 @@
+import { DataSource } from "apollo-datasource";
+import crypto from "crypto";
+
+export class Post {
+    constructor(data) {
+        this.id = crypto.randomBytes(16).toString("hex");
+        this.voters = new Map();
+        Object.assign(this, data);
+    }
+}
+export class User {
+    constructor(name) {
+        this.name = name;
+    }
+}
+
+export class InMemoryDataSource extends DataSource {
+    constructor() {
+        super();
+        this.posts = [];
+        this.users = [];
+    }
+
+    initialize({ context }) {}
+
+    allUsers() {
+        return Promise.resolve(this.users);
```

Comment:
Not sure why you implemented these methods as asynchronous.
Perhaps for forwards compatibility? I do expect actual DB connectors to have asynchronous APIs, though I don't know jack about Node.js..

Reply:
Yeah it was just to see and figure out how it would work with asynchronous responses :)